codegen/src/main/java/software/amazon/awssdk/codegen/poet/client/specs/VersionCompatibilityTestSpec.java

```diff
@@ -0,0 +1,99 @@
+import software.amazon.awssdk.codegen.poet.ClassSpec;
+import software.amazon.awssdk.codegen.poet.PoetUtils;
+
+public class VersionCompatibilityTestSpec implements ClassSpec {
+    private final IntermediateModel model;
+
+    public VersionCompatibilityTestSpec(IntermediateModel model) {
+        this.model = model;
+    }
+
+    @Override
+    public TypeSpec poetSpec() {
+        return PoetUtils.createClassBuilder(className())
+                        .addModifiers(Modifier.PUBLIC)
+                        .addMethod(compatibilityTest())
+                        .addMethod(isVersionCompatibleMethod())
+                        .build();
+    }
+
+    @Override
+    public ClassName className() {
+        return ClassName.get(model.getMetadata().getFullClientPackageName(), "VersionCompatibilityTest");
+    }
+
+    private MethodSpec compatibilityTest() {
+        ClassName serviceVersionInfo = ClassName.get(
+            model.getMetadata().getFullInternalPackageName(),
+            "ServiceVersionInfo"
+        );
+
+        ClassName versionInfo = ClassName.get("software.amazon.awssdk.core.util", "VersionInfo");
+        ClassName assertions = ClassName.get("org.assertj.core.api", "Assertions");
+
+        return MethodSpec.methodBuilder("checkCompatibility")
+                         .addModifiers(Modifier.PUBLIC)
+                         .addAnnotation(Test.class)
+                         .returns(void.class)
+                         .addStatement("String coreVersion = $T.SDK_VERSION", versionInfo)
+                         .addStatement("String serviceVersion = $T.VERSION", serviceVersionInfo)
+                         .addStatement("$T.assertThat(isVersionCompatible(coreVersion, serviceVersion))" +
+                                       ".withFailMessage(\"Core version %s must be equal to or newer than service version %s\", "
+                                       + "coreVersion, serviceVersion).isTrue()",
+                                       assertions)
+                         .build();
+    }
+
+    private MethodSpec isVersionCompatibleMethod() {
+        return MethodSpec.methodBuilder("isVersionCompatible")
+                         .addModifiers(Modifier.PRIVATE)
+                         .returns(boolean.class)
+                         .addParameter(String.class, "coreVersion")
+                         .addParameter(String.class, "serviceVersion")
+                         .addStatement("String normalizedCore = coreVersion.replace(\"-SNAPSHOT\", \"\")")
+                         .addStatement("String normalizedService = serviceVersion.replace(\"-SNAPSHOT\", \"\")")
+                         .addStatement("String[] coreParts = normalizedCore.split(\"\\\\.\")")
+                         .addStatement("String[] serviceParts = normalizedService.split(\"\\\\.\")")
+                         .addCode("\n")
+                         .addStatement("int coreMajor = Integer.parseInt(coreParts[0])")
```

Comment:
Are there ever cases we'd end up with non integer version parts?  I know in other SDKs we occasionally have like pre-release versions such as "2.0.0.pre".   Or are there ever times we would have less than 3 parts?